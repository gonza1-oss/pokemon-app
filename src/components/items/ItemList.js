import React from 'react';
import { Link } from "react-router-dom";

import getAllItems from '../../services/getAllItems';
import { getLoadingSpinner, getError } from '../../services/common';

function ItemList () {

  // graphql call to get the pokemon list
  const { loading, error, data } = getAllItems();
  if (loading) return getLoadingSpinner();
  if (error) return getError();

  // compone the pokemon card item
  const pokemons = data.pokemons.map(({ id, number, name, maxHP, maxCP, image, types }) => (
    <div className="ui card" key={id}>

      {/* pokemon image box */}
      <div className="image" style={{ background: `white url('${image}') no-repeat center`, backgroundSize: 'contain' }}></div>
      <div className="content">
        <Link to={`/${id}`} className="header">{name}</Link>
        <div className="meta">
          <span className="date">{number}</span>
        </div>

        {/* pokemon description box */}
        <div className="description">
          <p><b>Max HP:</b> {maxHP}</p>
          <p><b>Max CP:</b> {maxCP}</p>
          <p><b>Type:</b> {types.join(', ')}</p>
        </div>
      </div>

      {/* more information button */}
      <div className="ui bottom attached button">
        <Link to={`/${id}`}>
          <i className="ellipsis vertical icon"></i>
          More information
        </Link>
      </div>
    </div>
    ));

    return <div className="ui centered link cards">{pokemons}</div>;
}

export default ItemList;
