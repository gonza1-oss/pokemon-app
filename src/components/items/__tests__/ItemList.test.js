import React from 'react';
import ReactDOM from 'react-dom';
import { gql } from 'apollo-boost';
import { MockedProvider } from '@apollo/react-testing';
import { act, render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

import ItemList from '../ItemList';

const QUERY = gql`
{
  pokemons(first: 200) {
    id
    number
    name
    maxCP
    maxHP
    image
    types
  }
}`;

const MOCKS = [
  {
    request: {
      query: QUERY,
    },
    result: {
      data: {
        pokemons: [{
          id: "UG9rZW1vbjowMDE=",
          number: "001",
          name: "Bulbasaur",
          maxCP: 951,
          maxHP: 1071,
          image: "https://img.pokemondb.net/artwork/bulbasaur.jpg",
          types: [
            "Grass",
            "Poison"
          ]
        }],
      },
    },
  },
];

const MOCKS_ERROR = [
  {
    request: {
      query: QUERY,
    },
    result: null
  },
];

async function wait(ms = 0) {
  await act(() => {
    return new Promise(resolve => {
      setTimeout(resolve, ms);
    });
  });
}

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<MockedProvider mocks={[]}><ItemList /></MockedProvider>, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('renders loading spinner', async () => {
  const { container } = render(
    <BrowserRouter>
      <MockedProvider addTypename={false} mocks={MOCKS}>
        <ItemList />
      </MockedProvider>
    </BrowserRouter>
  );

  expect(container.textContent).toBe('Loading');

});

it('renders error', async () => {
  const { container } = render(
    <BrowserRouter>
      <MockedProvider addTypename={false} mocks={MOCKS_ERROR}>
        <ItemList />
      </MockedProvider>
    </BrowserRouter>
  );

  await wait(0);

  expect(container.textContent).toBe('Error :( ');

});

it('renders pokemon list', async () => {
  const { container } = render(
    <BrowserRouter>
      <MockedProvider addTypename={false} mocks={MOCKS}>
        <ItemList />
      </MockedProvider>
    </BrowserRouter>
  );

  await wait();

  expect(container.textContent).toMatchSnapshot();

  expect(container.textContent).toMatch('Bulbasaur');
  expect(container.textContent).toMatch('Type: Grass, Poison');
});
