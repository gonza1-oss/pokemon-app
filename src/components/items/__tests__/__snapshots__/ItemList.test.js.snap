// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders pokemon list 1`] = `"Bulbasaur001Max HP: 1071Max CP: 951Type: Grass, PoisonMore information"`;
